dmriprep/utils/grouping.py

```diff
@@ -0,0 +1,711 @@
+    >>> subject_data, layout = collect_data("separate_fmaps", SUBJECT_ID)
+    >>> group_by_warpspace(
+    ...     subject_data['dwi'], layout, True) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    [{'dwi_series': ['.../separate_fmaps/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz',
+       '.../separate_fmaps/sub-1/dwi/sub-1_dir-AP_run-2_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': 'j',
+      'concatenated_bids_name': 'sub-1_dir-AP'}]
+
+    Two DWI series, opposite PE directions, dedicated EPI fieldmap for each
+    >>> subject_data, layout = collect_data("mixed_fmaps", SUBJECT_ID)
+    >>> group_by_warpspace(
+    ...     subject_data['dwi'], layout, False) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    [{'dwi_series': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': 'epi',
+       'epi': ['.../mixed_fmaps/sub-1/fmap/sub-1_dir-PA_run-1_epi.nii.gz']},
+      'dwi_series_pedir': 'j',
+      'concatenated_bids_name': 'sub-1_dir-AP_run-1'},
+     {'dwi_series': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-PA_run-2_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': 'epi',
+       'epi': ['.../mixed_fmaps/sub-1/fmap/sub-1_dir-AP_run-2_epi.nii.gz']},
+       'dwi_series_pedir': 'j-',
+       'concatenated_bids_name': 'sub-1_dir-PA_run-2'}]
+
+    Same as last one, but ignore fieldmaps. The DWI series will be used for SDC instead
+    >>> subject_data, layout = collect_data("mixed_fmaps", SUBJECT_ID)
+    >>> group_by_warpspace(
+    ...     subject_data['dwi'], layout, True) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    [{'dwi_series': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': 'dwi',
+       'dwi': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-PA_run-2_dwi.nii.gz']},
+      'dwi_series_pedir': 'j',
+      'concatenated_bids_name': 'sub-1_dir-AP_run-1'},
+     {'dwi_series': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-PA_run-2_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': 'dwi',
+       'dwi': ['.../mixed_fmaps/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz']},
+      'dwi_series_pedir': 'j-',
+      'concatenated_bids_name': 'sub-1_dir-PA_run-2'}]
+
+
+    There is no metadata related to epi distortion: don't concatenate anything
+    >>> subject_data, layout = collect_data("missing_info", SUBJECT_ID)
+    >>> group_by_warpspace(
+    ...     subject_data['dwi'], layout, False) # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    [{'dwi_series': ['.../missing_info/sub-1/dwi/sub-1_dir-AP_run-1_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': '',
+      'concatenated_bids_name': 'sub-1_dir-AP_run-1'},
+     {'dwi_series': ['.../missing_info/sub-1/dwi/sub-1_dir-PA_run-2_dwi.nii.gz'],
+      'fieldmap_info': {'suffix': None},
+      'dwi_series_pedir': '',
+      'concatenated_bids_name': 'sub-1_dir-PA_run-2'}]
+
+    A bizarre mix of PE directions and some missing data
```

Comment:
This should be a comment

```suggestion
    >>> # A bizarre mix of PE directions and some missing data
```

same for all other instances (picked this one for no particular reason)